### DISCUSSION TITLE
How to manage Code Analysis rules ?

### DISCUSSION THREAD

**Discussion by irvnriir** [dotnet/roslyn#73545] [Q&A] [answered]
Task is to manage existing or add new rules, while keeping the default set of rules enabled by default .

The following seems to be supposed to work, but how to determine the AnalyzerId and Namespace by rule ID ?

Directory.Build.props
```
<Project>
	<CodeAnalysisRuleset>..\NETAnalyzers.ruleset</CodeAnalysisRuleset>
</Project>
```

NETAnalyzers.ruleset
```
<RuleSet Name="en" ToolsVersion="12.0">
	<Rules AnalyzerId=""
		RuleNamespace=""
	>		<!-- example: `AnalyzerId="Microsoft.Analyzers.ManagedCodeAnalysis" RuleNamespace="Microsoft.Rules.Managed"` -->
		<Rule Id="IDE1006" Action="Hidden" />
	</Rules>
</RuleSet>
```

Documentation is sparse, i also searched in this and https://github.com/dotnet/roslyn-analyzers repos and didn't find that .

**Comment by huoyaoyuan** [comment]
Ruleset is somehow legacy. Visual Studio has a designer for it and you can use as a reference.

The latest way is `analyerconfig`. You can see the documentation at https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/configuration-files#global-analyzerconfig and samples at https://github.com/dotnet/runtime/blob/main/eng/CodeAnalysis.src.globalconfig .

**Comment by irvnriir** [comment]
@huoyaoyuan ty for response. i did try this for extracting the reference data, but it doesn't contain a large set of items, specifically all that start with `IDE` .

i saw this. are default values for whether to run each custom rule can be specified in file which can be passed to project operations by `-analyzer:<<path to code analyzer>>` ? also, its important that C++ workflow uses rulesets.

**Reply by huoyaoyuan** [reply, marked as answer]
There are more documentation at https://learn.microsoft.com/visualstudio/code-quality/using-rule-sets-to-group-code-analysis-rules https://learn.microsoft.com/cpp/code-quality/using-rule-sets-to-specify-the-cpp-rules-to-run .
The IDE analyzers are integrated differently with CA analyzers. Try `EnforceCodeStyleOnBuild` MSBuild property may help.

**Reply by irvnriir** [reply]
I needed to open C# project and then open .ruleset, there isn't a global list. I saw that page, it doesn't include answer to that question, but i guess examination of the `Analyzer with code fix (.NET Standard)` template from ".NET Compiler Platform SDK" VS Component is needed. Though, the info or presets need to be in docs or on GitHub, not in a Windows-only app.
